### PR TITLE
meson build: fall back to dependancy if find_library fails

### PR DESCRIPTION
**Pull request by a-n-n-a-l-e-e**
find_library only does a basic search for libs where dependency can use cmake and pkg-config. this is helpful on systems like nix where libraries are located over may uniquely named directories

https://github.com/NixOS/nixpkgs/pull/268178

### CODE REVIEW

**Comment by bodono** [comment]
Thanks for doing this! It looks good to me, but I don't know anything about meson. Paging @enzbus (who is the expert here) to take a look.

File: meson.build [hunk 1]
```diff
@@ -7,17 +7,17 @@ blas_deps = []
...
+    blas_deps = [dependency('blas', static: get_option('link_blas_statically'), required : false)]
+    lapack_dep = dependency('lapack', static: get_option('link_blas_statically'), required : false)
     if lapack_dep.found()
         blas_deps += lapack_dep
     endif
-    cblas_dep = cc.find_library('cblas', static: get_option('link_blas_statically'), required : false)
+    cblas_dep = dependency('cblas', static: get_option('link_blas_statically'), required : false)
     if cblas_dep.found()
         blas_deps += cblas_dep
     endif
```
Comment (bodono) [inline meson.build:23]: Should we remove the cc.find_library on line 29 too?

File: meson.build [hunk 2]
```diff
@@ -7,17 +7,17 @@ blas_deps = []
 if host_machine.system() == 'darwin'
```
Comment (bodono) [inline meson.build:7]: Can we remove the 'cc = meson.get_compiler('c')' on line 5?

**Comment by enzbus** [comment]
This https://github.com/mesonbuild/meson/issues/6405#issuecomment-570772914 confirms OP statement. Probably all `find_library` can be switched. Does it pass the CI?

File: meson.build (hunk 2 above)
Comment (enzbus) [reply meson.build:7]: No we need it below

**Comment by enzbus** [comment]
It seems it fails on linux, we could try to do `apt install libopenblas-dev` since it doesn't pick the conda blas. Weird, looks like `find_library` is more robust.

File: meson.build (hunk 2 above)
Comment (bodono) [reply meson.build:7]: Oh right, missed even further below.

**Comment by bodono** [comment]
> It seems it fails on linux, we could try to do `apt install libopenblas-dev` since it doesn't pick the conda blas. Weird, looks like `find_library` is more robust.

Hmmmmm, that's strange. Not sure what the solution is here.

**Comment by enzbus** [comment]
> find_library only does a basic search for libs where dependency can use cmake and pkg-config. this is helpful on systems like nix where libraries are located over may uniquely named directories
> 
> 
> 
> https://github.com/NixOS/nixpkgs/pull/268178

@a-n-n-a-l-e-e do you have a clear usecase for this? it breaks our CI, you're right that what you're pointing is probably a more canonical way to do it but I'm not sure it's worth breaking what we have. This meson.build was adapted from numpy and scipy's.

**Comment by enzbus** [comment]
Also, you probably know this but this is only for building pip packages. scs proper has a separate build system, and repository.

**Comment by a-n-n-a-l-e-e** [comment]
> It seems it fails on linux, we could try to do `apt install libopenblas-dev` since it doesn't pick the conda blas. Weird, looks like `find_library` is more robust.

does seem that way -- i would've expect it to fall back on find_library, seems like the `openblas.pc`. scipy seems to use
`blas = dependency(['openblas', 'OpenBLAS'])`.

the use case i have to to work better on nix where the libraries are in unique directories.

**Comment by enzbus** [comment]
If the CI fails again (let's see) you could instead of replacing the find_library(), add an extra if else clause to also do dependency().

**Comment by a-n-n-a-l-e-e** [comment]
> If the CI fails again (let's see) you could instead of replacing the `find_library()`, add an extra `if else` clause to also do `dependency()`

looking at the docs dependancy defaults to auto detect, which is pkg-config + cmake + OSX thing. but there is a way to get it to also use system, which i think is  the same as find_library
https://mesonbuild.com/Dependencies.html#dependency-detection-method

let me see if i can get something that works. I appreciate you trying the CI.

File: meson.build (hunk 1 above)
Comment (enzbus) [reply meson.build:23]: @bodono I wouldn't touch the one on 29, it's very fragile it's the one that's made to work with the setup in CI to build windows wheels, it is not meant to be used in other settings

**Comment by a-n-n-a-l-e-e** [comment]
> If the CI fails again (let's see) you could instead of replacing the `find_library()`, add an extra `if else` clause to also do `dependency()`.

yeah -- this seems like the best option. unfortunately, the  `method:` arg for dependency only takes a string, not a list of options. will update soon -- but looks kind of gross. sorry.

File: meson.build (hunk 1 above)
Comment (bodono) [reply meson.build:35]: OK fair enough.

**Comment by bodono** [comment]
All checks pass! Is this ready to merge?

**Comment by a-n-n-a-l-e-e** [comment]
> All checks pass! Is this ready to merge?

yes

**Comment by a-n-n-a-l-e-e** [comment]
woot! thanks!

**Comment by a-n-n-a-l-e-e** [comment]
so -- it seems that openblas 0.3.22 renamed the pkg-config file from openblas.pc -> openblas64.pc.  so when checking for the openblas module probably want to have `dependancy(['openblas64', 'openblas', ...], ...)`. looking at the CI logs i can't tell what version of openblas it is building with, but if version 0.3.22 or later, could've been why my initial change was failing.

https://github.com/OpenMathLib/OpenBLAS/pull/3791

**Comment by bodono** [comment]
I presume openblas64 is using 64 bit integers? If so when linking against it we need to pass the compilation flag `BLAS64 = 1`.

**Comment by a-n-n-a-l-e-e** [comment]
> I presume openblas64 is using 64 bit integers? If so when linking against it we need to pass the compilation flag `BLAS64 = 1`.

thanks, i missed this. I don't think ILP64 is what we want. the nixos openblas build recently changed from LP64 to ILP64. investigating...

**Comment by a-n-n-a-l-e-e** [comment]
i was confused and thought the 64 indicated pointer width, then further confused with ILP64 that the  8 byte integers don't match the c. nixos didn't recently change from LP64 to ILP64, it is just that the meta data in the pkg-config file now indicates the 8 byte integer width.

for nixos -- the python-scs uses openblas LP64 (except for darwin, as that uses accelerate). that matches what numpy is using, unless there is a compelling reason to build for ILP64 matching numpy is simpler.

**Comment by enzbus** [comment]
Feel free to try with `openblas64` via `dependency`, if it passes the CI we can drop the initial `find_library` clauses which would simplify the code a little.

**Comment by a-n-n-a-l-e-e** [comment]
I don't think there is any pkg-config or cmake configuration data in the CI -- at least from looking at the files in the [openblas-devel]( https://packages.fedoraproject.org/pkgs/openblas/openblas-devel/fedora-rawhide.html#files) fedora package. in contrast with debian, ubuntu and nix which provide the configuration files to pkg-config and cmake
```
lib/.../cmake/openblas/OpenBLASConfig.cmake
lib/.../cmake/openblas/OpenBLASConfigVersion.cmake
lib/.../pkgconfig/blas-openblas.pc
lib/.../pkgconfig/lapack-openblas.pc
lib/.../pkgconfig/openblas.pc
```

i was mistaken to suggest openblas64.pc or that is for ILP64. LP64 uses openblas.pc

**Comment by enzbus** [comment]
OK so we keep the current (master branch) version, let us know if things work on your side.

**Pull request merged**